agent/internal/fluent.go

```diff
@@ -0,0 +1,263 @@
+	containers, err := c.ContainerList(context.Background(), types.ContainerListOptions{
+		Filters: filters.NewArgs(
+			filters.Arg("name", containerName),
+		),
+	})
+	if err != nil {
+		return 0, "", err
+	}
+	for _, cont := range containers {
+		ctx.Log().Infof("killing found Fluent container %s", cont.ID)
+		if err = c.ContainerKill(context.Background(), cont.ID, "KILL"); err != nil {
+			return 0, "", errors.Wrap(err, "failed to kill existing Fluent container")
+		}
+	}
+
+	// Pull the image.
+	_, _, err = c.ImageInspectWithRaw(context.Background(), imageName)
+	switch {
+	case err == nil:
+		// No error means the image is present; do nothing.
+	case client.IsErrNotFound(err):
+		// This error means the call to Docker went fine but the image doesn't exist; pull it.
+		pullResponse, pErr := c.ImagePull(context.Background(), imageName, types.ImagePullOptions{})
+		if pErr != nil {
+			return 0, "", pErr
+		}
+		if _, pErr = ioutil.ReadAll(pullResponse); pErr != nil {
+			return 0, "", pErr
+		}
+		if pErr = pullResponse.Close(); pErr != nil {
+			return 0, "", pErr
+		}
+	default:
+		// Something unexpected happened; propagate the error.
+		return 0, "", errors.Wrap(err, "failed to pull logging image")
+	}
+
+	fluentArgs, fluentFiles, err := a.fluentConfig()
+	if err != nil {
+		return 0, "", err
+	}
+
+	createResponse, err := c.ContainerCreate(
+		context.Background(),
+		&container.Config{
+			Image:        imageName,
+			Cmd:          fluentArgs,
+			ExposedPorts: nat.PortSet{exposedPort: struct{}{}},
+		},
+		&container.HostConfig{
+			AutoRemove: true,
+			PortBindings: nat.PortMap{
+				// A port of 0 makes Docker automatically assign a free port, which we read back below.
+				exposedPort: []nat.PortBinding{{HostIP: "127.0.0.1", HostPort: "0"}},
+			},
+		},
+		nil,
+		containerName,
+	)
+	if err != nil {
+		return 0, "", err
+	}
+
+	filesReader, _ := archive.ToIOReader(fluentFiles)
+
+	err = c.CopyToContainer(context.Background(),
+		createResponse.ID,
+		"/",
+		filesReader,
+		types.CopyToContainerOptions{},
+	)
+	if err != nil {
+		return 0, "", err
+	}
+
+	err = c.ContainerStart(context.Background(), createResponse.ID, types.ContainerStartOptions{})
+	if err != nil {
+		return 0, "", err
+	}
+
+	container, _ := c.ContainerInspect(context.Background(), createResponse.ID)
+	portStr := container.NetworkSettings.Ports[exposedPort][0].HostPort
+	port, _ := strconv.Atoi(portStr)
+	ctx.Log().Infof("Fluent listening on host port %d", port)
+
+	return port, createResponse.ID, nil
+}
+
+func (a *agent) killLoggingContainer() error {
+	c, err := client.NewClientWithOpts(client.FromEnv, client.WithVersion("1.40"))
+	if err != nil {
+		return errors.Wrap(err, "error connecting to Docker daemon")
+	}
+
+	defer func() {
+		if cErr := c.Close(); cErr != nil {
+			logrus.WithError(cErr).Warn("failed to close Docker connection")
+		}
+	}()
+
+	return c.ContainerKill(context.Background(), a.fluentContainerID, "KILL")
+}
```

Comment:
same comment here (and elsewhere) for pulling this docker machinery into a package. thoughts?